### PR TITLE
chore(client): remove dependency on server modules

### PR DESCRIPTION
**Pull request by vcrfxia**
### Description 

This PR is stacked on https://github.com/confluentinc/ksql/pull/5391 -- only the new commits (starting with "chore: introduce custom client exception type") need to be reviewed.

This PR removes the ksqldb-api-client module's non-test dependency on server modules (notably, `ksqldb-rest-app`). The three main changes as part of this effort were:
- introduction of KsqlClientException (rather than piggy-backing off KsqlRestClientException)
- introduction of KsqlException (rather than piggy-backing off KsqlApiException)
- removal of the client's JsonMapper in favor of the shared ApiJsonMapper

### Testing done 

Updated tests where appropriate.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-api-client/pom.xml [hunk 1]
```diff
@@ -54,6 +54,13 @@
 
         <!-- Required for running tests -->
 
+        <dependency>
+            <groupId>io.confluent.ksql</groupId>
```
Comment (vcrfxia) [inline ksqldb-api-client/pom.xml:58]: I'm not sure why this additional dependency is needed beyond the one below (with type `test-jar`). If someone with more Maven knowledge can explain why that'd be great 😄

**Comment by purplefox** [comment]
Looks great.

If you run mvn dependency tree on compile scope:

```
mvn dependency:tree -Dscope=compile
```

You will see that we are pulling in a lot of unwanted stuff via our ksqldb-common dependency (Kafka connect, Jetty etc). We should add dependency exclusions in the client pom to prevent these being dragged in.

**Comment by vcrfxia** [comment]
Thanks for the review @purplefox !

I've removed the Kafka Connect and Jetty dependencies that were being pulled in transitively from the ksqldb-rest-model dependency. Here's the new output of `mvn dependency:tree -Dscope=compile`:
```
[INFO] ----------------< io.confluent.ksql:ksqldb-api-client >-----------------
[INFO] Building ksqldb-api-client 6.0.0-SNAPSHOT                        [27/27]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:3.0.2:tree (default-cli) @ ksqldb-api-client ---
[INFO] io.confluent.ksql:ksqldb-api-client:jar:6.0.0-SNAPSHOT
[INFO] +- io.confluent.ksql:ksqldb-rest-model:jar:6.0.0-SNAPSHOT:compile
[INFO] |  +- io.confluent.ksql:ksqldb-common:jar:6.0.0-SNAPSHOT:compile
[INFO] |  |  +- io.confluent:kafka-connect-avro-converter:jar:6.0.0-SNAPSHOT:compile
[INFO] |  |  |  \- io.confluent:kafka-connect-avro-data:jar:6.0.0-SNAPSHOT:compile
[INFO] |  |  +- io.confluent:common-logging:jar:6.0.0-SNAPSHOT:compile
[INFO] |  |  +- org.apache.kafka:kafka-streams:jar:6.0.0-ccs-SNAPSHOT:compile
[INFO] |  |  |  \- org.rocksdb:rocksdbjni:jar:5.18.4:compile
[INFO] |  |  \- org.apache.commons:commons-lang3:jar:3.3.1:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.10.2:compile
[INFO] |     \- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.2:compile
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.10.2:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-core:jar:2.10.2:compile
[INFO] +- io.vertx:vertx-core:jar:3.9.0:compile
[INFO] |  +- io.netty:netty-common:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-buffer:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-transport:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-handler:jar:4.1.48.Final:compile
[INFO] |  |  \- io.netty:netty-codec:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-handler-proxy:jar:4.1.48.Final:compile
[INFO] |  |  \- io.netty:netty-codec-socks:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-codec-http:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-codec-http2:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-resolver:jar:4.1.48.Final:compile
[INFO] |  \- io.netty:netty-resolver-dns:jar:4.1.48.Final:compile
[INFO] |     \- io.netty:netty-codec-dns:jar:4.1.48.Final:compile
[INFO] +- org.reactivestreams:reactive-streams:jar:1.0.3:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.26:compile
[INFO] +- org.slf4j:slf4j-log4j12:jar:1.7.26:compile
[INFO] |  \- log4j:log4j:jar:1.2.17:compile
[INFO] +- io.confluent:common-utils:jar:6.0.0-SNAPSHOT:compile
[INFO] \- io.confluent:assembly-plugin-boilerplate:zip:resources:6.0.0-SNAPSHOT:provided
[INFO] ------------------------------------------------------------------------
```
There's still transitive dependencies on `kafka-connect-avro-converter` and `kafka-streams` from the ksqldb-common dependency. Do you think we should try removing these too? I don't have a sense of how much worthwhile it is to continue pruning dependencies.

Also, do you know why these are among the only four transitive dependencies listed from ksqldb-common? ksqldb-common has a bunch more dependencies and I don't see them marked as optional or anything else that would cause not all of them to appear in the output of `mvn dependency:tree -Dscope=compile`.

**Comment by purplefox** [comment]
I think we definitely need to get rid of the avro, streams and rocksdb jars too - they could be pretty big!

Now sure why all the deps aren't showing  - I did notice that guava has an exclusion in the parent pom... not sure if that is related. Also guava is declared as a test dep in the common pom.xml - seems fishy! Does it appear if you do `mvn dependency:tree -Dscope=test`

**Comment by purplefox** [comment]
Also, I think you can exclude transitive dependencies using wildcards. So instead of excluding everything one by one, I think you can exclude * , then _include_ only the things we want.

Haven't tried it though!

https://stackoverflow.com/questions/35929936/how-to-turn-off-transitive-dependencies-for-maven-projects

**Comment by vcrfxia** [comment]
OK, here are the new dependencies after I removed the transitive dependencies on Kafka Connect and Kafka Streams:
```
[INFO] ----------------< io.confluent.ksql:ksqldb-api-client >-----------------
[INFO] Building ksqldb-api-client 6.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:3.0.2:tree (default-cli) @ ksqldb-api-client ---
[INFO] io.confluent.ksql:ksqldb-api-client:jar:6.0.0-SNAPSHOT
[INFO] +- io.confluent.ksql:ksqldb-rest-model:jar:6.0.0-SNAPSHOT:compile
[INFO] +- io.confluent.ksql:ksqldb-common:jar:6.0.0-SNAPSHOT:compile
[INFO] +- com.google.guava:guava:jar:24.1.1-jre:compile
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] |  +- org.checkerframework:checker-compat-qual:jar:2.0.0:compile
[INFO] |  +- com.google.errorprone:error_prone_annotations:jar:2.1.3:compile
[INFO] |  +- com.google.j2objc:j2objc-annotations:jar:1.1:compile
[INFO] |  \- org.codehaus.mojo:animal-sniffer-annotations:jar:1.14:compile
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.10.2:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.10.2:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.10.2:compile
[INFO] |     \- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.2:compile
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-guava:jar:2.10.2:compile
[INFO] +- io.vertx:vertx-core:jar:3.9.0:compile
[INFO] |  +- io.netty:netty-common:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-buffer:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-transport:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-handler:jar:4.1.48.Final:compile
[INFO] |  |  \- io.netty:netty-codec:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-handler-proxy:jar:4.1.48.Final:compile
[INFO] |  |  \- io.netty:netty-codec-socks:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-codec-http:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-codec-http2:jar:4.1.48.Final:compile
[INFO] |  +- io.netty:netty-resolver:jar:4.1.48.Final:compile
[INFO] |  \- io.netty:netty-resolver-dns:jar:4.1.48.Final:compile
[INFO] |     \- io.netty:netty-codec-dns:jar:4.1.48.Final:compile
[INFO] +- io.vertx:vertx-codegen:jar:3.9.0:compile
[INFO] |  \- org.mvel:mvel2:jar:2.3.1.Final:compile
[INFO] +- org.reactivestreams:reactive-streams:jar:1.0.3:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.26:compile
[INFO] +- org.slf4j:slf4j-log4j12:jar:1.7.26:compile
[INFO] |  \- log4j:log4j:jar:1.2.17:compile
[INFO] +- io.confluent:common-utils:jar:6.0.0-SNAPSHOT:compile
[INFO] \- io.confluent:assembly-plugin-boilerplate:zip:resources:6.0.0-SNAPSHOT:provided
[INFO] ------------------------------------------------------------------------
```

I had to reintroduced the custom JsonMapper since ApiJsonMapper depends on 
```
<dependency>
    <groupId>org.apache.kafka</groupId>
    <artifactId>connect-json</artifactId>
</dependency>
```
for JsonConverter and 
```
<dependency>
    <groupId>io.confluent.ksql</groupId>
    <artifactId>ksqldb-udf</artifactId>
</dependency>
```
for SqlType.

**Comment by vcrfxia** [comment]
> Now sure why all the deps aren't showing - I did notice that guava has an exclusion in the parent pom... not sure if that is related. Also guava is declared as a test dep in the common pom.xml - seems fishy!

> Does it appear if you do mvn dependency:tree -Dscope=test?

No, they don't. `mvn dependency:tree -Dscope=test` showed the same limited set of transitive dependencies from ksqldb-common as `mvn dependency:tree -Dscope=compile`.

**Comment by vcrfxia** [comment]
One thing that concerns me with removing transitive dependencies like this is that our integration tests have diverged from the actual user experience, since there are a bunch of test-scoped dependencies that can allow integration tests to pass whereas a user app that just pulls in the `ksqldb-api-client` module might fail. Do you have advice for how to plug this test gap @purplefox ?

I saw this happen with the missing JsonConverter and SqlType transitive dependencies when I initially excluded transitive deps without reintroducing the custom JsonMapper -- the integration tests were passing but a simulated user app I wrote failed.

**Comment by purplefox** [comment]
I think we need to remove guava too. Do we really need this in the client? (Does it build and do tests pass if we exclude it?).

Guava is one of those libraries we really don't want in a client - it's pretty widely used so the chance of a dependency clash when a user uses the client in their own app and they already use a different version of Guava is pretty high.

**Comment by purplefox** [comment]
> One thing that concerns me with removing transitive dependencies like this is that our integration tests have diverged from the actual user experience, since there are a bunch of test-scoped dependencies that can allow integration tests to pass whereas a user app that just pulls in the ksqldb-api-client module might fail. Do you have advice for how to plug this test gap @purplefox ?

Could you elaborate? Not sure I understand.

**Comment by vcrfxia** [comment]
> > One thing that concerns me with removing transitive dependencies like this is that our integration tests have diverged from the actual user experience, since there are a bunch of test-scoped dependencies that can allow integration tests to pass whereas a user app that just pulls in the ksqldb-api-client module might fail. Do you have advice for how to plug this test gap @purplefox ?
> 
> Could you elaborate? Not sure I understand.

Here's an example change illustrating the test gap: https://github.com/vcrfxia/ksql/commit/fb7d11bbfdad48752dc49d33eef20249b6dbbfc9
On that branch, integration tests (including all the ones in `ClientIntegrationTest.java`) pass, but running `ExampleApp.java` (which simulates a client app that only has the ksqldb-api-client module dependency) fails with `java.lang.NoClassDefFoundError: org/apache/kafka/connect/json/JsonConverter`. This shows that we have a test gap.

**Comment by purplefox** [comment]
This seems like an issue with the way dependencies are set up for our tests - most likely the test deps bring in all the server side stuff too so the jars are still available even though the client doesn't depend on them.

This can probably be resolved by creating a custom surefire run:

https://maven.apache.org/surefire/maven-surefire-plugin/examples/configuring-classpath.html

And excluding the test scope classpath, and adding in the compile classpath instead. (Or something like that!)

**Pull request merged**